Groupify v1.0.0: Better STI/polymorphism and various fixes

joelvh opened the pull request:
This has turned into an extensive refactor to improve STI/polymorphism. It also attempts to DRY up some aspects of the code.

- Drops Rails 4.0 and 4.1 support
- Adds `has_group` and `has_member` class methods to configure associations (for STI)
- Adds ability for polymorphic groups and members with `polymorphic_groups` and `polymorphic_members`
- Adds ability to customize default `groups` and `members` association names
- Adds ability to customize default member class
- Adds `membership_types_for_*` method on groups and group members to get list of membership types
- Adds support for multiple membership types when querying using `as(membership_types)` or helper methods like `user.in_any_group?(group1, group2, as: membership_types)` (using SQL `IN(...)`/`OR` query)
- Updated logic when appending a group or member to the related association.
- Utilize dynamic modules to add common logic to groups and group members (ActiveRecord)
- Fixes an issue of duplicate membership records if no membership type is specified.
- Fixes properly clear association caches after `merge!` on groups and `delete` or `destroy` on associations
- Fixes additional bugs for cases that were not covered (see new tests in test suite)

Please provide feedback.

Thanks!

joelvh committed:
Add the ability to specify a map of association names to class names

joelvh committed:
Introduce `has_member` with `:class_name` option for custom class on association

joelvh committed:
Override `<<` on collections to properly add membership_type

joelvh committed:
Add member to members collection rather than the group to groups collection for polymorphic groups compatibility

joelvh committed:
Don't set default, for backwards-compatibility

joelvh committed:
Query based group class (to support subclasses)

joelvh committed:
Remove unused option

joelvh committed:
Clear association cache after deleting associated models

joelvh committed:
Fix query to be extensible

joelvh committed:
Refactored to mimic `<<` method returning `false` if an invalid argument is given otherwise returns collection

joelvh committed:
Add to collection with options (e.g. membership type)

joelvh committed:
Specify group class name when merging in case it's a subclass

joelvh committed:
Remove `group_type` based on group class because we handle STI separately

joelvh committed:
Make it easier to add subclassed group associations with STI

joelvh committed:
Use ActiveRecord `base_class` in case member is STI

joelvh committed:
Updated docs for `has_group`

joelvh committed:
Revert back to using association and assume STI

joelvh committed:
Added ability for `group.add` to throw a validation exception

joelvh committed:
Remove unneeded `default_members_association_name` feature

dwbutler commented:
This looks promising, but I shall want to see tests in order to have confidence that the solution here works.

Also, it seems that `validate!` wasn't added until Rails 4.2, but this gem supports 4.0+.

joelvh committed:
Don't use `validate!` to ensure Rails 4.0.x compatibility

joelvh committed:
Move extensions modules to separate files

joelvh committed:
Consolidate queries for deleting records

joelvh committed:
Refactor to handle groups and members associations with same logic

joelvh committed:
Moved `delete` and `destroy` methods to shared module

joelvh committed:
Move finder methods to appropriate modules

joelvh committed:
Reuse delete/destroy logic

joelvh committed:
Require modules

joelvh committed:
Mark methods protected

joelvh committed:
Define `add` explicitly as raising an exception instead of using `opts`

joelvh committed:
Alias `add` and `<<` in shared module

joelvh committed:
Simplify abstractions and fix some reference errors

joelvh committed:
Fixed alias methods

joelvh committed:
Fix `super` reference

joelvh committed:
Added `group_type` to query to properly build membership record

joelvh commented:
@dwbutler changed out the `validate!` method for some plain Ruby. Also ended up refactoring the association extension modules to share the repeated logic, so that consistency can be more closely maintained.

What are some specific tests you would want to see?

joelvh committed:
Added tests to make sure inverse associations are updated after `delete` and `destroy`

joelvh committed:
Added tests to make sure members and groups aren't added more than once

joelvh committed:
Added tests to check subclassing

joelvh commented:
@dwbutler added some tests. Some direction on what tests to add is welcome. Thanks.

joelvh commented:
@dwbutler Wanted to see if you had a chance to look at this? Thanks.

dwbutler commented:
@joelvh Sorry been a crazy few weeks. Definitely planning to get to this soon.

joelvh commented:
@dwbutler got it. Will stand by.

joelvh commented:
hi @dwbutler saw you were able to make some updates recently. Did you get a chance to look at these updates?

joelvh committed:
Fix whitespace

dwbutler commented:
@joelvh Yes, I've been going through it bit by bit when I have time. It's taking a while because you have effectively rewritten large parts of the gem! But I like what I'm seeing so far.

dwbutler commented on lib/groupify/adapter/active_record/group.rb:
```diff
@@ -50,7 +44,10 @@ def merge!(source)
-          member.groups
+          #member.groups
+          joins(:group_memberships_as_group).
+          merge(Groupify.group_membership_klass.where(member_id: member.id, member_type: member.class.base_class.to_s)).
```
@joelvh I couldn't quite understand what this change was for, because when I revert it back the tests still pass. From the commits, it looks like this adds support for querying on subclasses so that STI can be supported. This is a good example of where a specific test for this would be helpful.

I think this line can be simplified by changing to `merge(member.group_memberships_as_member)`

joelvh replied:
@dwbutler I refactored the queries to use `merge` instead of `where`, as you started doing. Tests still pass. As you indicated, there are probably very specific STI-related scenarios we need to create some tests for. I'll see what I can come up with as I use this in our project, or if this use of `merge` handles things for us.

dwbutler commented:
@joelvh I'd like to help iterate on this with you so we can get this merged. It would be helpful if you and I could both commit to the same branch. I could give you access to the official Groupify repo, or you can give me access to your fork.

Also, since you've made such deep changes to Groupify, it would be great to have your help as a maintainer, if you're open to it.

dwbutler commented on lib/groupify/adapter/active_record/association_extensions.rb:
```diff
@@ -0,0 +1,100 @@
+
+      def as(membership_type)
+        return self unless membership_type
+        where(group_memberships: {membership_type: membership_type})
```
Per https://github.com/dwbutler/groupify/pull/58 this needs to be changed to `merge(Groupify.group_membership_klass.as(membership_type))` to support a custom group membership model.

joelvh replied:
@dwbutler fixed - must have been missed in the rebase

joelvh committed:
Use `merge` instead of `where`

joelvh committed:
Refactor `where` with `merge`

joelvh commented:
@dwbutler for simplicity, I've added you to https://github.com/identification-io/groupify/ for now. I'd be happy to help you maintain this project too.

joelvh commented on lib/groupify/adapter/active_record/association_extensions.rb:
```diff
@@ -0,0 +1,100 @@
+      end
+
+      def delete(*records)
+        remove_children_from_parent(records.flatten, :delete){ |*args| super(*args) }
```
@dwbutler I don't really like how `super` is implemented here, but was having problems with getting a reference to `super` otherwise (e.g. `remove_children_from_parent(records.flatten, :delete, &method(:super))` or something like that didn't work... newer versions of Ruby allow `super_method`, but didn't want to go that route)

joelvh committed:
Let Rails generate the queries on `group_id` and `group_type`

joelvh committed:
Update `group_type` for completeness

joelvh committed:
Remove commented code

joelvh committed:
Remove duplicated extension modules from rebase

joelvh commented:
@dwbutler it seems changing `where(group_id: group.id)` to `where(group: group)` works in Rails 4.1+ with SQLite. Making that change makes the query automatically also query on `group_type` column. Otherwise, the queries referring to `group_id` should also add the `group_type` criteria for completeness.

What are your thoughts on requiring Rails 4.1+ with SQLite?

joelvh committed:
Simplify member criteria

joelvh committed:
helpers to query on polymorphic groups and members

joelvh committed:
Use merge and helpers for query

joelvh committed:
clearer logic

joelvh commented:
@dwbutler I added a helper method to generate a query where the groups could be different types. However, I'm realizing that that's meant more for polymorphism, not STI.

joelvh committed:
Add ability to call `has_group :custom_groups, "CustomGroup"`

joelvh committed:
Clarify logic and simplify some steps

joelvh committed:
Pass each record as separate parameter

joelvh committed:
Include type column when counting matches

joelvh committed:
Removed unused method

joelvh committed:
Added helper method to make it easier to read SQL criteria

joelvh committed:
DRY logic

joelvh committed:
Simplify logic when finding matching named group

joelvh committed:
Simplify variable use

joelvh committed:
Simplify membership query building

joelvh committed:
Code spacing

joelvh committed:
Refer to `arel_table` consistently

dwbutler commented:
@joelvh Sorry for the delay (again). I'm okay with supporting Rails 4.1+ since Rails 4.0 will soon no longer be supported.

Better support for polymorphism would be amazing! Of course, the more that is changed in this PR, the more time it will take to review and the more tests there should be.

I think the scale of these changes might even warrant a major (1.x) release.

joelvh committed:
Removed unused methods

joelvh committed:
Formatted for clarity

joelvh commented:
Hi @dwbutler - I'm happy to help with creating more tests. My primary focus has been to make sure that the existing tests pass and to create new ones where I found bugs previously.

I'd appreciate any direction the tests you might have, as it's probably getting to very granular cases quickly. I will have to see about creating some tests for STI and polymorphism.

dwbutler commented:
@joelvh I'm mainly interested in seeing tests for STI and polymorphism (there are some tests but they are currently pretty low coverage); and for multiple group classes (since that's a major new feature).

joelvh committed:
Consolidate association extensions and move helpers to `ActiveRecord` module

joelvh committed:
Fix polymorphic adding of members to groups by bypassing association reference

joelvh committed:
Added `memberships_merge` helper to named groups

joelvh committed:
Add helper method to infer class and association names

joelvh committed:
Clean up `membership_type` usage and check presence better/minimally

joelvh committed:
Improve inferring parent model when model can be group and member

joelvh committed:
Consolidate `membership_merge` logic

joelvh committed:
Exclude `ActiveRecord::AssociationTypeMismatch` test because it is not raised with polymorphic logic

joelvh commented:
@dwbutler added a test for testing the polymorphic logic. Revealed 2 things:

* Had to add members to groups by creating membership records directly rather than adding to `member.groups` association because that default association only supports one type of model. Not going through these associations does not cause a problem. This also means a test was disabled related to expecting a type exception.
* There is logic to infer "parent" and "child" relationships (which comes into play to DRY the code when either adding groups to a member or adding members to a group). This had to be updated when a model can be both a group and a user -- you can't infer and need a flag specified indicating whether the group or the member is the parent.

You're right -- this is turning into a larger release version. Tests pass. Let me know if you think the polymorphic ones added are sufficient.

joelvh committed:
Build query via chain instead of control logic

joelvh committed:
Set blank values to nil

joelvh committed:
Throw "type mismatch" exception when adding to association

joelvh committed:
Update Mongoid with some similar code cleanup as ActiveRecord

joelvh commented:
@dwbutler So, it's possible to create memberships with groups or members from different tables,  but there's not really easy way to get a list of all groups for a member. I'm thinking about a helper method that would somehow get a list of all groups from the group memberships. However, it would return an array instead of an ActiveRecord::Relation, which could then be extended.

Will probably just ignore this for now.

Thoughts?

joelvh committed:
Fixed filter fallback in tests

joelvh committed:
Fix flattening arrays

joelvh committed:
DRY up Mongoid similar to ActiveRecord

joelvh committed:
Fix Mongoid test

joelvh committed:
Rails 4.0 doesn't like merging empty hash, so we build the query step by step

joelvh committed:
Fix test to use specific IDs

joelvh committed:
Clear association cache on child after creating membership

joelvh committed:
Fix tests for Postgres which changes the order of records

joelvh committed:
Fix tests to compare arrays properly

joelvh committed:
Simplify methods by removing outdated aliasing

joelvh committed:
Add test to make sure associations are reset after deletion

joelvh committed:
Add default`members` association in more intuitive spot

dwbutler committed:
Fix named group support in Rails 5

Rails 5 made `belongs_to` associations required by default. When using named groups, the group association is expected to be empty. This fixes the issue in a Rails 4 compatible way by adding `required: false`

Fixes https://github.com/dwbutler/groupify/issues/65

joelvh committed:
Clean up `has_member` and `has_group` and allow association options

joelvh committed:
Fix wrong variable name reference

joelvh committed:
Add `has_groups` and `has_group` to Mongoid

joelvh committed:
Add proper default class

joelvh committed:
Only look up base class if needed

dwbutler committed:
Fix Rails 4.0 - 4.1, which don't support the `required` option

Also, specs for models that are only named group members, and don't use groups

joelvh committed:
Added `Groupify.ignore_base_class_inference_errors` to swallow inference errors with `base_class`

joelvh committed:
Use `source_type` instead of `class_name`

joelvh committed:
Consolidate `has_member` and `has_group` and add error handling

joelvh committed:
Use base class

joelvh committed:
Update tests with `autoload` to help resolve circular class dependencies

joelvh committed:
Moved test classes to separate files for autoloading

joelvh commented:
@dwbutler so, adding tests lead to some enhancements for polymorphism and STI. However, the rabbit hole lead to some circular dependency resolution issues in the tests. Specifically, when the `has_member` and `has_group` are called, they attempt to get the base class of the inferred class name for each association. However, in some cases, this causes a reference to a class that wasn't defined yet. I added `autoload` to the test suite, and it seems to have resolved things.

However, now I'm trying to understand an error that is breaking the tests:

`Association named 'users' was not found on Organization; perhaps you misspelled it?`

Any idea what this might stem from?

joelvh committed:
Indicate value of variable better in exception (e.g. when nil)

joelvh commented:
@dwbutler I meant to say that this problem does not happen in a Rails app, only in the tests.

joelvh commented:
@dwbutler OK - so, the issue is that, while the parent class `Group` is being loaded, the Groupify associations are created. In the tests, the group has a `users` association. When that loads the `User` class, it builds the `organizations` association. `Organization` inherits from `Group`, which means `Organization is inheriting from a class that is still in the midst of loading and is not fully loaded. It seems that `Organization` only inherits from `Group` what `Group` has loaded so far, and doesn't inherit the rest when `Group` is finished loading. This whole thing is like a recursive loading process that gets confusing and seems to screw things up outside of Rails.

Do you think we can implement autoloading like Rails does, so this is not an issue?

Alternatively, I have to revisit the code that "constantizes" strings and tries to get the base class while configuring Groupify. If there were a way to evaluate these things in some "lazy" way, that would work too. 

Thoughts?

joelvh committed:
Simplify

joelvh committed:
Add option to return a string version of inferred class name rather than trying to constantize

joelvh committed:
Fix variable name

joelvh committed:
Add `class_name` in case the default group class is a STI subclass

joelvh committed:
Use default group class when base class can't be resolved

joelvh committed:
Make default `members` and `groups` association names configurable

joelvh committed:
Abstract out extension methods

joelvh committed:
Introduce `PolymorphicChildren` collection class to

joelvh committed:
Disabled autoloading in tests (for now)

joelvh committed:
Remove unused argument

joelvh committed:
Split out polymorphic classes for multiple uses

joelvh committed:
Properly merge queries

joelvh committed:
Allow modifying query

joelvh committed:
Crate a relation

joelvh committed:
Scope properly

joelvh committed:
Fix variable name

joelvh committed:
Fix query nesting

joelvh committed:
Fix type

joelvh committed:
Add `pretty_print`

joelvh commented:
@dwbutler I've reverted tests because autoloading wasn't working as expected. I've implemented some fallbacks that allow tests to pass as they did before.

I also implemented a "polymorphic collection", which is a proxy to `GroupMembership` queries and returns either the related `group` or `member` properties. The goal is to retrieve all groups and members, regardless of the related table. Using `includes` to avoid N+1.

All logic goes through `user.polymorphic_groups` and `group.polymorphic_members` internally. It also mimics an `ActiveRecord::Relation` for adding/removing records and could be used as a replacement for `user.groups` and `group.members`. However, I felt that the way people may be using this is more extensive than the tests we have and wanted to introduce minimal breakage.

I realize this PR keeps growing, but I think it's nearing completion.

Let me know what you think.

Thanks.

joelvh committed:
Rename block argument for consistency

joelvh committed:
Use `instance_eval` for consistency

joelvh committed:
Fix up some consistency in the code

joelvh committed:
Remove duplicated class definitions

joelvh committed:
Fix postgres by changing from `GROUP BY` to `DISTINCT ON`

joelvh committed:
Added test to make sure polymorphic groups are unique

joelvh committed:
Make sure that `count` is correct based on PostgreSQL DISTINCT vs. GROUP BY

joelvh committed:
Make sure memberships are added properly based on role

joelvh committed:
Moved parent/child logic into `ParentProxy` class to consolidate and simplify

joelvh committed:
Fix for group deletion deleting group membership after merge... not sure why this is happening now - no logic changes.

joelvh committed:
Simplify class

joelvh committed:
Move query logic to `ParentQueryBuilder` class

joelvh committed:
Clarify logic

joelvh committed:
DRY up `has_many` creation

joelvh committed:
Rename "query" to "scope"

joelvh committed:
determine child type

joelvh committed:
Fix Rails 4.0-5.0 tests using `extending` on model class when chaining scopes

joelvh committed:
Fix DISTINCT queries in PostgreSQL for SELECT vs COUNT

joelvh committed:
Drop support for Rails 4.0

joelvh committed:
Use helper to build query

joelvh committed:
DRY up code with helpers

joelvh committed:
Make default member class configurable

joelvh committed:
Fix `merge` to use `all` scope for Rails 4.0-5.0

joelvh commented:
@dwbutler all tests are passing. Introduced a couple new classes for polymorphism. This should be done now and ready for review and testing by you and others. Let me know what you think.

joelvh committed:
Make public method names more user friendly

joelvh committed:
Fix `children_association` parent reference

joelvh committed:
Add `superclass` helper to get inherited values for STI

joelvh committed:
Use inherited class and association names

joelvh committed:
Standardize same options on group members as on groups

joelvh committed:
DRY configuring defaults

joelvh committed:
Return nothing if child association name not specified

joelvh committed:
Don't delegate methods to create records

joelvh committed:
Updated README with polymorphic and STI documentation

joelvh committed:
Added tests for disabling default associations

joelvh committed:
Add block to extend `has_many`

joelvh committed:
Indicate 4.1+ support

joelvh committed:
Updated README

joelvh committed:
Revert ActiveRecord configuration to inline for clarity

joelvh committed:
Replicated configuration setup from ActiveRecord to Mongoid

joelvh commented:
@dwbutler do you use Mongoid? It's a bit tricky to maintain that adapter when not really using it myself and relying only on tests.

joelvh committed:
Renamed "merge" methods for clarity

joelvh committed:
Removed redundant scope merge

joelvh committed:
Disable `appraisal` gem because it causes segfaults - loosen other gem versions

joelvh committed:
Don't add default associations by default, but provide `configure_legacy_defaults!` for backwards-compatibility

joelvh committed:
Updated README

joelvh committed:
Consistent naming of `opts`

joelvh committed:
Cleanup unused code

joelvh committed:
Update generator with default configuration comments

joelvh committed:
Note that groups and members need to be persisted first

joelvh committed:
Add method to get membership types for a group/member combo

joelvh committed:
Add `ParentQueryBuilder` methods as extension methods to models

joelvh committed:
Build extensions in separate files for clarity

joelvh committed:
Generate extension modules rather than using helper classes

joelvh committed:
Pass extension block to `has_many`

joelvh committed:
Continued de-duplication of code with dynamic module creation

joelvh committed:
Renamed `ModelMembershipExtensions` to `ModelExtensions`

joelvh committed:
Remove references to old classes

joelvh committed:
Fix module naming

joelvh committed:
Added ability to search on multiple membership types (SQL `OR`/`IN(...)`) query

joelvh committed:
Fix tests and DSL to allow both group and member implemented on one model

joelvh committed:
Set `class_name` when inferring class from association name (STI)

joelvh committed:
Removed Rails 4.1 support

joelvh committed:
Removed unnecessary option

joelvh committed:
Fixes Rails 4.2 bug merging associations

joelvh committed:
Adding should throw validation exception because old version called `create!`

joelvh committed:
Fix tests to use specific order of records for PostgreSQL

joelvh committed:
Fix merging relations that don't go through group memberships

joelvh committed:
Fix test to order comparison for PostgreSQL

joelvh committed:
Remove db type check

joelvh committed:
Use new helper method

dwbutler commented:
@joelvh Sorry for the delayed response. I'll take another look over this.

I do not use Mongoid any more. I've been considering splitting this into two separate gems, but it's not urgent.

joelvh commented:
@dwbutler ok, look forward to your feedback.

joelvh committed:
Check for relation to generate subquery

joelvh committed:
Fixed exclusion path

joelvh committed:
Require newer versions of gems

joelvh committed:
Simplify generated variable names and clarify attribute references

joelvh committed:
Fix test that is using a class that wasn't designated as a group member

joelvh committed:
Fix inverse association references for unsaved records

joelvh committed:
Test for persistence of new records when adding to groups

joelvh committed:
Added brief descriptions to polymorphic classes

joelvh committed:
Fix MySQL GROUP BY error when sql_mode=only_full_group_by

Fixed error by removing GROUP BY all together because selecting the ID
column without being listed in the GROUP BY clause is not allowed in
MySQL when in “strict” mode.

joelvh committed:
Add instructions on how to run tests

joelvh committed:
Made further cross-database SQL compatibility fixes for DISTINCT

joelvh committed:
Add line info for dynamic modules

joelvh committed:
Merge remote-tracking branch 'dwbutler/fix/rails5-named-groups' into feature/polymorphic_adding

joelvh committed:
Clean membership handling - could add multiple memberships at once

joelvh committed:
Update authors

joelvh commented:
@dwbutler wanted to see if you had any more time to review this?

joelvh committed:
Fix attempt to join and merge to do so on group model, not group membership

joelvh committed:
Join and merge is flawed - removed

joelvh committed:
Clarify error source

joelvh commented:
@dwbutler checking in on your review of this. Anything I can do to help make the review easier for you?

dwbutler commented:
Hey @joelvh thanks for following up. Going through some personal stuff right now so I haven't had time to spend on open source like I usually do.

I pretty much have full trust in you at this point. So when I have time, I'll figure out how to give you full access to the project so you can merge and release it.

joelvh commented:
@dwbutler sorry to hear that. I'm happy to help maintain. I'll try to get a list of devs who might be interested in testing a Release Candidate, unless you have a list of people that might be interested.

dwbutler commented:
@joelvh Sorry for the long long delay. You should have full access now to this repository and to push the gem on rubygems. Welcome aboard!

joelvh committed:
Don't infer association or class name if `:class_name` option is specified

rposborne commented:
Would love to see this merged. Anything I can do to help?

dwbutler commented:
@rposborne Would you be interested in helping maintain Groupify as well?

rposborne commented:
@dwbutler I don't have much time to offer, I do offer "gem gardening" though, If you just need someone else with the keys to make sure those who are eager have a point of contact.  But I cannot take on any real responsibility :/

tomekr commented:
would be great to see this merged! Curious if you ended up getting access @joelvh ?

monroemann commented:
Is this gem still being monitored and updated?

dwbutler commented:
I am no longer maintaining this gem. I don't write Ruby any more and don't use this gem, so it's difficult for me to stay involved in its maintenance. If anyone is interested in maintaining it, I would recommend forking it to a new organization. I can facilitate this if anyone is willing to take on the mantle.
